src/chat.cpp

```diff
@@ -474,22 +501,36 @@ void ChatPrompt::input(const std::wstring &str)
 
 void ChatPrompt::addToHistory(const std::wstring &line)
 {
+	std::wstring old_line = getLine();
+	if (m_history_index < m_history.size()) {
+		HistoryEntry &entry = m_history[m_history_index];
+		if (entry.saved && entry.line == line) {
+			entry.line = *entry.saved;
+			entry.saved = nullopt;
+			// Remove the entry if it is now a duplicate
+			if (std::find(m_history.begin() + m_history_index + 1, m_history.end(), entry)
```

Comment:
You could make this calculation safer by using an iterator in place of `HistoryEntry &entry = m_history[m_history_index];`.